File: games/billiards/server/rules/placement.ts
import { GameRuleError } from "@tabletop/game-sdk/server";

import { tableSpecFor } from "../../shared/table.js";
import type { BilliardsBall } from "../../shared/view.js";
import type { BallInHandZone, BilliardsMatchState } from "../state.js";

const POSITION_EPSILON = 1e-9;

export type CuePlacementValidation =
  | { readonly ok: true }
  | {
      readonly ok: false;
      readonly ruleCode:
        | "CUE_OUT_OF_BOUNDS"
        | "CUE_OUTSIDE_D"
        | "CUE_OUTSIDE_BEHIND_LINE"
        | "CUE_IN_POCKET"
        | "CUE_OVERLAPS_BALL";
    };

export function checkCuePlacement(
  state: Readonly<BilliardsMatchState>,
  x: number,
  y: number,
  zone: BallInHandZone = state.ballInHandZone ?? "anywhere",
): CuePlacementValidation {
  const table = tableSpecFor(state.settings.mode);
  const radius = table.ballDiameter / 2;
  if (
    !Number.isFinite(x) ||
    !Number.isFinite(y) ||
    x < radius - POSITION_EPSILON ||
    x > table.width - radius + POSITION_EPSILON ||
    y < radius - POSITION_EPSILON ||
    y > table.height - radius + POSITION_EPSILON
  ) {
    return { ok: false, ruleCode: "CUE_OUT_OF_BOUNDS" };
  }

  if (zone === "d") {
    if (table.baulkLineX === null || table.dRadius === null) {
      return { ok: false, ruleCode: "CUE_OUTSIDE_D" };
    }
    const dx = x - table.baulkLineX;
    const dy = y - table.height / 2;
    if (
      x > table.baulkLineX + POSITION_EPSILON ||
      dx * dx + dy * dy > table.dRadius * table.dRadius + POSITION_EPSILON
    ) {
      return { ok: false, ruleCode: "CUE_OUTSIDE_D" };
    }
  }

  if (zone === "behind-line") {
    if (table.baulkLineX === null || x > table.baulkLineX + POSITION_EPSILON) {
      return { ok: false, ruleCode: "CUE_OUTSIDE_BEHIND_LINE" };
    }
  }

  if (
    table.pockets.some(
      ({ x: pocketX, y: pocketY, captureRadius }) =>
        Math.hypot(x - pocketX, y - pocketY) < captureRadius - POSITION_EPSILON,
    )
  ) {
    return { ok: false, ruleCode: "CUE_IN_POCKET" };
  }

  if (
    state.balls.some(
      (ball) =>
        ball.kind !== "cue" &&
        !ball.pocketed &&
        Math.hypot(x - ball.x, y - ball.y) < table.ballDiameter - POSITION_EPSILON,
    )
  ) {
    return { ok: false, ruleCode: "CUE_OVERLAPS_BALL" };
  }
  return { ok: true };
}

export function validateCuePlacement(
  state: Readonly<BilliardsMatchState>,
  x: number,
  y: number,
  zone: BallInHandZone = state.ballInHandZone ?? "anywhere",
): void {
  const result = checkCuePlacement(state, x, y, zone);
  if (!result.ok) throw new GameRuleError(result.ruleCode, { x, y, zone });
}

export function placeCueBall(
  balls: readonly BilliardsBall[],
  x: number,
  y: number,
): readonly BilliardsBall[] {
  let found = false;
  const placed = balls.map((ball) => {
    if (ball.kind !== "cue") return ball;
    found = true;
    return { ...ball, pocketed: false, rotation: 0, x, y };
  });
  if (!found) throw new TypeError("Billiards state is missing the cue ball");
  return placed;
}

File: games/billiards/native/src/rules/placement.rs
use super::error::{RuleError, RuleResult};
use super::types::{BallInHandZone, BilliardsBall, BilliardsBallKind, BilliardsMatchState};
use crate::geometry::table_spec;

const POSITION_EPSILON: f64 = 1e-9;

pub fn check_cue_placement(
    state: &BilliardsMatchState,
    x: f64,
    y: f64,
    zone: Option<BallInHandZone>,
) -> RuleResult<()> {
    let zone = zone
        .or(state.ball_in_hand_zone)
        .unwrap_or(BallInHandZone::Anywhere);
    let table = table_spec(state.settings.mode);
    let radius = table.ball_diameter / 2.0;
    if !x.is_finite()
        || !y.is_finite()
        || x < radius - POSITION_EPSILON
        || x > table.width - radius + POSITION_EPSILON
        || y < radius - POSITION_EPSILON
        || y > table.height - radius + POSITION_EPSILON
    {
        return Err(RuleError::rule(
            "CUE_OUT_OF_BOUNDS",
            "Cue-ball placement is outside the playable surface",
        ));
    }

    if zone == BallInHandZone::D {
        let Some(baulk_line_x) = table.baulk_line_x else {
            return Err(RuleError::rule(
                "CUE_OUTSIDE_D",
                "The selected table has no D",
            ));
        };
        let Some(d_radius) = table.d_radius else {
            return Err(RuleError::rule(
                "CUE_OUTSIDE_D",
                "The selected table has no D",
            ));
        };
        let dx = x - baulk_line_x;
        let dy = y - table.height / 2.0;
        if x > baulk_line_x + POSITION_EPSILON
            || dx * dx + dy * dy > d_radius * d_radius + POSITION_EPSILON
        {
            return Err(RuleError::rule(
                "CUE_OUTSIDE_D",
                "Cue-ball placement must be inside the D",
            ));
        }
    }

    if zone == BallInHandZone::BehindLine
        && table
            .baulk_line_x
            .is_none_or(|baulk_line_x| x > baulk_line_x + POSITION_EPSILON)
    {
        return Err(RuleError::rule(
            "CUE_OUTSIDE_BEHIND_LINE",
            "Cue-ball placement must be behind the baulk line",
        ));
    }

    if table.pockets.iter().any(|pocket| {
        (x - pocket.capture_x).hypot(y - pocket.capture_y)
            < pocket.capture_radius - POSITION_EPSILON
    }) {
        return Err(RuleError::rule(
            "CUE_IN_POCKET",
            "Cue-ball placement overlaps a pocket capture area",
        ));
    }

    if state.balls.iter().any(|ball| {
        ball.kind != BilliardsBallKind::Cue
            && !ball.pocketed
            && (x - ball.x).hypot(y - ball.y) < table.ball_diameter - POSITION_EPSILON
    }) {
        return Err(RuleError::rule(
            "CUE_OVERLAPS_BALL",
            "Cue-ball placement overlaps an object ball",
        ));
    }
    Ok(())
}

pub fn place_cue_ball(balls: &[BilliardsBall], x: f64, y: f64) -> RuleResult<Vec<BilliardsBall>> {
    let mut found = false;
    let placed = balls
        .iter()
        .cloned()
        .map(|mut ball| {
            if ball.kind == BilliardsBallKind::Cue {
                found = true;
                ball.pocketed = false;
                ball.rotation = 0.0;
                ball.x = x;
                ball.y = y;
            }
            ball
        })
        .collect();
    if !found {
        return Err(RuleError::invalid(
            "CUE_BALL_MISSING",
            "Billiards state is missing the cue ball",
        ));
    }
    Ok(placed)
}
